Sources/Plasma/FeatureLib/pfConsole/pfConsole.cpp

```diff
@@ -693,21 +626,39 @@ void    pfConsole::IHandleKey( plKeyEventMsg *msg )
             }
         }
         // or are they just typing in a working line
-        else if( strlen( fWorkingLine ) < kMaxCharsWide - 2 && key != 0 )
+        else
         {
-            for( i = strlen( fWorkingLine ) + 1; i > fWorkingCursor; i-- )
-                fWorkingLine[ i ] = fWorkingLine[ i - 1 ];
-
-            // TODO: What about keys outside of `char` range?
-            fWorkingLine[fWorkingCursor++] = char(key);
-
+            // FIXME: make the console unicode friendly.
+            IHandleCharacter((char)key);
             findAgain = false;
             findCounter = 0;
             IUpdateTooltip();
         }
     }
 }
 
+void    pfConsole::IHandleCharacter(const char c)
+{
+    // Properly handle embedded newlines as execution triggers.
+    if (c == '\r')
+        return;
+    if (c == '\n') {
+        IExecuteWorkingLine();
+        return;
+    }
+
+    size_t i = strnlen(fWorkingLine, std::size(fWorkingLine));
+
+    // The current working line is too long. Ignore this character.
+    if (!(i < kMaxCharsWide - 2 && c != '\0'))
```

Comment:
Super minor quibble, but I feel like De-Morgan-ing this makes it clearer and match the comment better.
```suggestion
    if (i >= kMaxCharsWide - 2 || c == '\0'))
```